### DISCUSSION TITLE
Correct way to probe whether a terminal supports the kitty keyboard protocol?

### DISCUSSION THREAD

**Discussion by elebow** [kovidgoyal/kitty#6080] [Q&A] [answered]
https://sw.kovidgoyal.net/kitty/keyboard-protocol/#detection-of-support-for-this-protocol says the program should emit `CSI ? u` (query kitty mode flags) and `CSI c` (query primary device attributes). If the terminal responds with the kitty mode and then the primary device attributes, then it supports the kitty keyboard protocol.

I think I misunderstand how the terminal will "respond" to the progam. In both of the Python scripts below, the terminal prints the expected response sequences in the terminal window. I can see the sequences, but the program cannot read them from stdin (or in the second example, the `/dev/tty` file handle).

What am I doing wrong?

`kitty 0.26.5`, if it matters.

```python
import sys

sys.stdout.write("[?u") # query kitty progressive enhancement flags
sys.stdout.write("[c") # query primary device attributes
sys.stdout.flush
while True:
    ch = sys.stdin.read(1)
    if ch == "u":
        print("kitty protocol supported")
        break
    elif ch == "c":
        print("kitty protocol not supported")
        break
```

```python

tty = open("/dev/tty", "r+")
tty.write("[?u") # query kitty progressive enhancement flags
tty.write("[c") # query primary device attributes
tty.flush()
while True:
    ch = tty.read(1)
    print(ch)
    if ch == "u":
        print("kitty protocol supported")
        break
    elif ch == "c":
        print("kitty protocol not supported")
        break
  ```

**Comment by page-down** [comment]
> ... how the terminal will "respond" to the progam ...

There is nothing special here, after sending the escape sequence you will be able to read the reply from stdin.
You can see the terminal printed on the screen because ECHO is not turned off.

https://docs.python.org/3/library/termios.html

> ... cannot read them from stdin ...

I noticed this: `sys.stdout.flush` .

**Reply by elebow** [reply]
ECHO explains why I see the response, but I still can't read it from stdin. Flushing stdout shouldn't affect that.

Here's a simpler script:
```python
import sys

sys.stdout.write("[?u")
sys.stdout.write("[c")
response = sys.stdin.read(3)
print(f"got response -{response}-")
```

It blocks at `sys.stdin.read(3)`, and pressing `ctrl d` causes:
```
got response --
^[[?0u^[[?62;c
```

**Comment by kovidgoyal** [comment, marked as answer]
You need to turn off line bufferring on stdin.

**Reply by elebow** [reply]
That was it; thank you.
